### ISSUE TITLE
ComboBox "filtering" is not putting back valid items, v8.0.5

### ISSUE THREAD

**Issue by krasnyt** [vaadin/framework#9027]
Code to reproduce:

```
@Override
protected void init(VaadinRequest request) {

    ComboBox<String> cb = new ComboBox<>("cp");
    cb.setDataProvider(new ListDataProvider<String>(asList("ffoo", "mmoo", "ppoo")));

    setContent(cb);
}
```

Steps to reproduce:
1) start typing "ff"
2) dropdown list only contains "ffoo"
3) click outside so the combobox loses focus (and therefore deletes "ff")
    4a) **expected:** see again all 3 items when opening dropdown
    4b) **real:** I see only "ffoo" altough "ff" has been from ComboBox deleted

**Comment by limburgie** [comment]
It's a shame that this issue is not solved yet, as this concerns basic usage of the ComboBox component. Check also this issue https://github.com/vaadin/framework/issues/7790, which describes the exact same problem. No response from Vaadin since August 2016.

The strange thing is, if you try this behavior on the Vaadin Sampler (https://demo.vaadin.com/sampler/#ui/data-input/multiple-value/combo-box) you can't reproduce the error, although you see a notable delay between reopening the combobox with the wrong amount of items and seeing the option list fill again with all items.

Does anyone have a workaround that can "reset" the ComboBox when it is reopened after it was filtered using the keyboard?

**Comment by bebopCL** [comment]
Hi,
I'm struggling with the same problem in my Vaadin 8 application, but I can reproduce this behavior in the Vaadin sampler:
[https://demo.vaadin.com/sampler/#ui/data-input/multiple-value/combo-box](https://demo.vaadin.com/sampler/#ui/data-input/multiple-value/combo-box)

1) Click into the textfield of the combobox
2) Type something e.g. "ALB"
 -> popup will show two Items [ALBANIA] and [SVALBARD AND JAN MAYEN]
3) Click next to the combobox
-> combobox will loose the focus
-> hint shows that 'no country is selected'
4) Click on the popup button of the combobox again
-> There are still only 2 items visible / filter is not resetted

OR

1) Click into the textfield of the combobox
2) Type something e.g. "ALB"
 -> popup will show two Items [ALBANIA] and [SVALBARD AND JAN MAYEN]
3) Select one of the items
-> Value changed hint is shown on the right bottom of the page
4) Click next to the combobox
-> combobox will loose the focus
-> combobox field shows the flag an name of the selected item
4) Click on the popup button of the combobox again
-> There are still only 2 items visible / filter is not resetted

Tested with 
chome 57.0
IE 11.0
Firefox 52.0

Like limburgie said: the popup is refreshed with a little delay on reopening the combobox and shows all items again, but only if it does **not** loose the focus after typing the filtering-text.

**Comment by tsuoanttila** [comment]
This was picked to 8.0.7

**Comment by OsamaFelFel** [comment]
The issue still exists in 8.1.6

**Comment by mifaber** [comment]
The issue is also still there in 8.3.0...
Does anyone have some ideas for a workaround?

**Comment by TatuLund** [comment]
Have you checked 8.3.1, which has this fix: https://github.com/vaadin/framework/issues/10284 ? There will be further improvement in 8.4.0 via https://github.com/vaadin/framework/pull/10606

**Comment by mifaber** [comment]
Hi, okay, i just checked 8.3.1 and 8.3.2, but it is still there.
The problem is, that the filter is not resetted, after i reopen the window. So the DataProvider is has still the filter from the previous usage of the screen...
The topic is also discussed above, but it seems that with [#9381](https://github.com/vaadin/framework/pull/9381) the reset of "lastFilter" in the ClientSide is removed again. 

Do you see any kind of workaround on that?

**Comment by TatuLund** [comment]
Another fix, that may impact this under works is: https://github.com/vaadin/framework/pull/10796

**Comment by mifaber** [comment]
Okay, this sounds interesting. Is there a planned release date for this fix?

**Comment by krasnyt** [comment]
Still not fixed in 8.4.0.

**Comment by tsuoanttila** [comment]
Please move the discussion from this old already closed to #9915 

The remaining issue is a race condition with the filter changing while lazy loading being too slow and is being investigated.
